File: python/balance_parens.py
def balance_parens(string):
    #creating base list to mutate
    char_list = list(string)
    #creating a stack to push open bracket indexes into for matching
    stack = []
    #looping through list for matching
    for i, char in enumerate(char_list):
        if char == ')':
            #if stack is populated
            if stack:
                # means there is a match so I want to pop the index from stack and continue
                stack.pop()
            #if nothing in stack, no match so replace the unmatch close with nothing
            else:
                char_list[i] = ''
        #if character is an open bracket I want to append to the stack
        elif char == '(':
            stack.append(i)
    #looping over stack, if there are contents, means there are unmatched open brackets
    #so I want to replace them with nothing in the list
    for idx in stack:
        char_list[idx] = ''
    #finally returning the mutated list joined back into a string
    return ''.join(char_list)

File: js/balanceParens.js
balanceParens = (string) => {
    //splitting input string to have a mutable array
    const charArray = string.split('');
    //creating a stack to push and pop from
    const stack = [];
    //looping over array for matching
    for (let i = 0; i < charArray.length; i++) {
        if (charArray[i] === ')') {
            //as long as there is something in stack meaning there is a matching open
            if (stack.length > 0) {
                //popping off the value in the stack
                stack.pop();
            //if there is no corresponding opening, replce the closing bracket with nothing  
            } else {
                charArray[i] = '';
            }
        //if element is an open bracket I want to push the index into the stack
        } else if (charArray[i] === '(') {
            stack.push(i);
        }
    }
    //looping through stack accounting for any left over non-matching open brackets
    //and replacing them with nothing
    for (let j = 0; j < stack.length; j++) {
        charArray[stack[j]] = ''
    }
    //returning the mutated array joined back into a string
    return charArray.join('')
}
module.exports = { balanceParens }
